### ISSUE TITLE
Publish to jnbinder

### ISSUE THREAD

**Issue by BoPeng** [vatlab/jnbinder#4]
I plan to implement a general `sos publish` (And `%publish` magic) to publish sos notebooks to repositories. The notebook would have some mechanism for keywords and meta data, and the receipient (server) would have some RESTful API to accept the metadata and notebook itself. The command itself would have plugins to publish to different services. Not sure if this would be something useful for jnbinder.

**Comment by gaow** [comment]
Okey, so the `repository` you are talking about is not just `git`, but something customized that resides on sos-notebooks.com? (You mentioned different services).

Not sure if I'm getting your point fully, but ultimately there are 2 types of servers I believe: one that serves notebooks that users can actually play with, the other only displays them (static HTML). `jnbinder` certainly fall into the 2nd category. What it does is to simply take a bunch of notebooks, and return a website just like the SoS website. So I can imaging a mechanism that sends notebooks to a server where `jnbinder` (or generically a software you wrote) converts it to a website and publish -- of course it altomatically builds a site with index pages or tags by keywords etc. 

So, 

1. Do you agree with my 2 type of server assumption?
2. Given the 2nd type of server is easy to implement, is something like `jnbinder` a software that you have in mind that converts submitted notebooks into organized websites for display?

**Comment by BoPeng** [comment]
1. Yes, but how to publish the accepted notebook is something that the server will worry about, not sos notebook.
2. Yes. The key here is each server will implement some API that accepts remote submission. 

What I am trying to do is
1. SoS defines interface to keyword storage and retrieval, and calls plugins to send notebooks to external website.
2. Server defines API to acept notebooks with required meta information, and store (convert first if needed) them in their own way.
3. Each server also provides a plugin to the publish command to define meta information that it needs and how to talk to the API.

Here server can be jnbinder, some github repo (in this case no dedicated server is needed and the plugin will only be used to enforce things such as meta information), some project metadata registration service that our department is trying to enforce, and sos-notebooks of course.

**Comment by gaow** [comment]
I see. Yes it seems my understanding is on the same page as your proposal. All these are crucial steps to make it happen.

> and store (convert first if needed) them in their own way

This is where the current version of `jnbinder` fits in -- convert the notebook to a website along with other notebooks, in its own way. Hopefully it can be used as a prototype for this bit, if we take a modular approach at this? Of course it should be tailored for the specific purpose in dealing with the meta-info, but I can see advantages to make this step modular for more flexibility.

**Comment by BoPeng** [comment]
Yes, I believe`%publish` would remove a major bottleneck for users to publish their notebooks. It is counter productive to log in to a website, manually enter meta data and upload notebooks.

**Comment by gaow** [comment]
When I simply put everything from:

https://github.com/vatlab/jnbinder

to this folder:

https://github.com/vatlab/sos-docs/tree/master/src

and edit `config.yml` file a bit, then run `./release`, I get:

[sos-docs-src.zip](https://github.com/vatlab/jnbinder/files/1746746/sos-docs-src.zip)

As you can see `jnbinder` basically streamlines codes that `sos-docs` wiki uses, to publish to HTML.  Other than putting things together, I guess the couple of major changes I made are:

1. automatically create index pages. The index genreated here look somewhat not polished because sos-docs did not exactly follow the convention jnbinder requires
2. a simple mechanism to password protect some pages

It works great for sharing my own work. It would be great if you find it good starting point to build on these codes to publish, although I suspect you need something a lot more advanced.
